README.rst
cURL as DSL
===================

This command generates source code from cURL command line options.

WebSite: http://shibukawa.github.io/curl_as_dsl/

Install
---------

.. code-block:: bash

   $ go get -u github.com/shibukawa/curl_as_dsl

Usage
-------

.. code-block:: bash

   $ curl_as_dsl [global options] curl [curl options]

Global Options
~~~~~~~~~~~~~~~~~~~~~~~~

.. code-block:: none

   -t, --target     Code generator name. Now it supports the following generators:

       go, golang         : Golang      (net/http)
       py, python         : Python 3    (http.client)
       node, js.node      : Node.js     (http.request)
       xhr, js.xhr        : Browser     (XMLHttpRequest)
       java               : Java        (java.net.HttpURLConnection)
       objc, objc.session : Objective-C (NSURLSession)
       objc.connection    :             (NSURLConnection)
       vim                : Vim script  (WebAPI-vim)

Supported cURL Options
~~~~~~~~~~~~~~~~~~~~~~~~

.. code-block:: none

   [usage]

          --basic                             Use HTTP Basic Authentication (H)
          --compressed                        Request compressed response (using deflate or gzip)
      -d, --data=DATA                         HTTP POST data (H)
          --data-ascii=DATA                   HTTP POST ASCII data (H)
          --data-binary=DATA                  HTTP POST binary data (H)
          --data-urlencode=DATA               HTTP POST data url encoded (H)
      -G, --get                               Send the -d data with a HTTP GET (H)
      -F, --form=KEY=VALUE                    Specify HTTP multipart POST data (H)
          --form-string=KEY=VALUE             Specify HTTP multipart POST data (H)
      -H, --header=LINE                       Pass custom header LINE to server (H)
      -I, --head                              Show document info only
      -x, --proxy=[PROTOCOL://]HOST[:PORT]    Use proxy on given port
      -e, --referer=                          Referer URL (H)
      -X, --request=COMMAND                   Specify request command to use
          --tr-encoding                       Request compressed transfer encoding (H)
      -T, --upload-file=FILE                  Transfer FILE to destination
          --url=URL                           URL to work with
      -u, --user=USER[:PASSWORD]              Server user and password
      -A, --user-agent=STRING                 User-Agent to send to server (H)

License
---------

MIT License




_doc/history.rst
History
============

11/25/2015
-------------

* Fix Golang post form (``-F``) issue (thanks `spiegel-im-spiegel <http://qiita.com/spiegel-im-spiegel/items/042751d98e315e4e3382>`_).

11/06/2015
--------------

* ``--http2`` is supported (only for Golang and node.js)
* ``--insecure`` is supported (only for Golang and node.js)

03/14/2015
--------------

* Initial version


run_test_go.sh
#!/bin/bash

rm test/*
set -e
echo "case 1: simple get"
./httpgen curl http://localhost:18888 > test/test.go
pushd test;go build;./test;popd

echo "case 2: simple post with data"
./httpgen curl -d test http://localhost:18888 > test/test.go
pushd test;go build;./test;popd

echo "case 3: post multiple datas"
./httpgen curl -d test -d hello http://localhost:18888 > test/test.go
pushd test;go build;./test;popd

echo "case 4: post url encoded data"
./httpgen curl --data-urlencode="test% =" http://localhost:18888 > test/test.go
pushd test;go build;./test;popd

echo "case 5: get with parameter"
./httpgen curl -G -d hello http://localhost:18888 > test/test.go
pushd test;go build;./test;popd

echo "case 6: get with aprameter (key=value style)"
./httpgen curl -G -d hello=world http://localhost:18888 > test/test.go
pushd test;go build;./test;popd

echo "case 7: post with aprameter (key=value style)"
./httpgen curl -X POST -G -d hello=world http://localhost:18888 > test/test.go
pushd test;go build;./test;popd

echo "case 8: simple post without data"
./httpgen curl -X POST http://localhost:18888 > test/test.go
pushd test;go build;./test;popd

echo "case 9: simple post with local file content"
./httpgen curl -X POST -T test.go http://localhost:18888 > test/test.go
pushd test;go build;./test;popd

echo "case 10: post form"
./httpgen curl -F hello=world http://localhost:18888 > test/test.go
pushd test;go build;./test;popd

echo "case 10-2: post form (2)"
./httpgen curl -F hello=world -F good=morning http://localhost:18888 > test/test.go
pushd test;go build;./test;popd

echo "case 11: post text data from local file"
./httpgen curl --data-ascii @test.go http://localhost:18888 > test/test.go
pushd test;go build;./test;popd

echo "case 12: post text data from local files"
./httpgen curl --data-ascii @test.go --data-ascii @test.go http://localhost:18888 > test/test.go
pushd test;go build;./test;popd

echo "case 13: post data from local file"
./httpgen curl --data-binary @test.go http://localhost:18888 > test/test.go
pushd test;go build;./test;popd

echo "case 14: post data from local files"
./httpgen curl --data-binary @test.go --data-binary @test.go http://localhost:18888 > test/test.go
pushd test;go build;./test;popd

echo "case 15: post url encoded data from local file"
./httpgen curl --data-urlencode @test.go http://localhost:18888 > test/test.go
pushd test;go build;./test;popd

echo "case 16: post url encoded data from local files"
./httpgen curl --data-urlencode @test.go --data-urlencode @test.go http://localhost:18888 > test/test.go
pushd test;go build;./test;popd

echo "case 17: send file in form protocol"
./httpgen curl -F "file=@test.go" http://localhost:18888 > test/test.go
pushd test;go build;./test;popd

echo "case 18: send file in form protocol with explicit name and type"
./httpgen curl -F "file=@test.go;filename=nameinpost;type=text/plain" http://localhost:18888 > test/test.go
pushd test;go build;./test;popd

echo "case 19: send file in form protocol (2)"
./httpgen curl -F "file=<test.go" http://localhost:18888 > test/test.go
pushd test;go build;./test;popd

echo "case 20: send file in form protocol with explicit type (2)"
./httpgen curl -F "file=<test.go;type=text/plain" http://localhost:18888 > test/test.go
pushd test;go build;./test;popd

echo "case 21: (Full Client) get with aprameter and header"
./httpgen curl -H "Accept: text/html" -G -d hello=world http://localhost:18888 > test/test.go
pushd test;go build;./test;popd

echo "case 22: (Full Client) simple post with data and user-agent"
./httpgen curl --user-agent="Netscape 4.7" -d test http://localhost:18888 > test/test.go
pushd test;go build;./test;popd

echo "case 23: post url encoded data from local files with compressed option"
./httpgen curl --compressed --data-urlencode @test.go --data-urlencode @test.go http://localhost:18888 > test/test.go
pushd test;go build;./test;popd

echo "case 24: Basic authentication"
./httpgen curl -u USER:PASS http://localhost:18888 > test/test.go
pushd test;go build;./test;popd

echo "case 25: simple get with https"
./httpgen curl --insecure https://localhost:18889 > test/test.go
pushd test;go build;./test;popd

echo "case 26: simple get with http2"
./httpgen curl --insecure --http2 https://localhost:18889 > test/test.go
pushd test;go build;./test;popd

